File: typescript/api-ts/src/storage.ts
/**
 * S3/MinIO storage operations for the Gallery feature.
 */
import {
  CreateBucketCommand,
  DeleteObjectCommand,
  HeadBucketCommand,
  ListObjectsV2Command,
  PutBucketPolicyCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { minioConfig } from "./config.js";

let s3Client: S3Client | null = null;

/** Get S3 client for MinIO (singleton). Returns null if not configured. */
export function getS3Client(): S3Client | null {
  if (s3Client) {
    return s3Client;
  }

  const { endpoint, accessKey, secretKey } = minioConfig;
  if (!endpoint || !accessKey || !secretKey) {
    return null;
  }

  s3Client = new S3Client({
    endpoint,
    region: "us-east-1",
    credentials: {
      accessKeyId: accessKey,
      secretAccessKey: secretKey,
    },
    forcePathStyle: true,
  });

  return s3Client;
}

/** Create bucket if it doesn't exist and set public read policy. */
export async function ensureBucket(s3: S3Client): Promise<void> {
  const { bucket } = minioConfig;

  try {
    await s3.send(new HeadBucketCommand({ Bucket: bucket }));
  } catch (error: unknown) {
    const err = error as { name?: string; $metadata?: { httpStatusCode?: number } };
    if (err.name === "NotFound" || err.$metadata?.httpStatusCode === 404) {
      await s3.send(new CreateBucketCommand({ Bucket: bucket }));
    }
  }

  // Ensure public read policy
  try {
    const publicPolicy = {
      Version: "2012-10-17",
      Statement: [
        {
          Sid: "PublicRead",
          Effect: "Allow",
          Principal: "*",
          Action: ["s3:GetObject"],
          Resource: [`arn:aws:s3:::${bucket}/*`],
        },
      ],
    };
    await s3.send(
      new PutBucketPolicyCommand({
        Bucket: bucket,
        Policy: JSON.stringify(publicPolicy),
      }),
    );
  } catch {
    // Policy might already be set
  }
}

export interface GalleryImage {
  key: string;
  url: string;
  lastModified?: string;
  size?: number;
}

/** List all images in the gallery bucket, sorted by last modified (newest first). */
export async function listImages(s3: S3Client): Promise<GalleryImage[]> {
  const { bucket, publicBaseUrl, endpoint } = minioConfig;
  const images: GalleryImage[] = [];
  let continuationToken: string | undefined;

  do {
    const response = await s3.send(
      new ListObjectsV2Command({
        Bucket: bucket,
        ContinuationToken: continuationToken,
      }),
    );

    for (const item of response.Contents || []) {
      if (
        item.Key &&
        (item.Key.endsWith(".jpg") || item.Key.endsWith(".jpeg") || item.Key.endsWith(".png"))
      ) {
        const baseUrl = publicBaseUrl || endpoint;
        images.push({
          key: item.Key,
          url: `${baseUrl?.replace(/\/$/, "")}/${bucket}/${item.Key}`,
          lastModified: item.LastModified?.toISOString(),
          size: item.Size,
        });
      }
    }

    continuationToken = response.IsTruncated ? response.NextContinuationToken : undefined;
  } while (continuationToken);

  // Sort by lastModified descending
  images.sort((a, b) => {
    if (!a.lastModified || !b.lastModified) {
      return 0;
    }
    return new Date(b.lastModified).getTime() - new Date(a.lastModified).getTime();
  });

  return images;
}

/** Delete an image from the gallery bucket. */
export async function deleteImage(s3: S3Client, key: string): Promise<void> {
  await s3.send(new DeleteObjectCommand({ Bucket: minioConfig.bucket, Key: key }));
}

File: python/api-python/storage.py
"""S3/MinIO storage operations for the Gallery feature."""

import json

import boto3
from botocore.exceptions import ClientError
from config import GalleryImage, minio_config

# Singleton S3 client
_s3_client = None


def get_s3_client():
    """
    Get S3 client for MinIO (singleton).
    Returns None if MinIO is not configured.
    """
    global _s3_client

    if _s3_client is not None:
        return _s3_client

    if not minio_config.is_configured:
        return None

    _s3_client = boto3.client(
        "s3",
        endpoint_url=minio_config.endpoint,
        aws_access_key_id=minio_config.access_key,
        aws_secret_access_key=minio_config.secret_key,
        region_name="us-east-1",
    )

    return _s3_client


def ensure_bucket(s3_client) -> None:
    """Create bucket if it doesn't exist and set public read policy."""
    bucket = minio_config.bucket

    # Check if bucket exists
    try:
        s3_client.head_bucket(Bucket=bucket)
    except ClientError as e:
        error_code = e.response.get("Error", {}).get("Code")
        if error_code in ("404", "NoSuchBucket"):
            s3_client.create_bucket(Bucket=bucket)
        else:
            raise

    # Set public read policy
    try:
        public_policy = {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Sid": "PublicRead",
                    "Effect": "Allow",
                    "Principal": "*",
                    "Action": ["s3:GetObject"],
                    "Resource": [f"arn:aws:s3:::{bucket}/*"],
                }
            ],
        }
        s3_client.put_bucket_policy(Bucket=bucket, Policy=json.dumps(public_policy))
    except ClientError:
        # Policy might already be set
        pass


def list_images(s3_client) -> list[GalleryImage]:
    """List all images in the gallery bucket, sorted by last modified (newest first)."""
    bucket = minio_config.bucket
    public_base_url = minio_config.public_base_url or minio_config.endpoint
    images: list[GalleryImage] = []

    paginator = s3_client.get_paginator("list_objects_v2")

    for page in paginator.paginate(Bucket=bucket):
        for item in page.get("Contents", []):
            key = item.get("Key", "")
            if key.endswith((".jpg", ".jpeg", ".png")):
                base_url = public_base_url.rstrip("/")
                images.append(
                    GalleryImage(
                        key=key,
                        url=f"{base_url}/{bucket}/{key}",
                        last_modified=item.get("LastModified").isoformat()
                        if item.get("LastModified")
                        else None,
                        size=item.get("Size"),
                    )
                )

    # Sort by last_modified descending (newest first)
    images.sort(key=lambda x: x.last_modified or "", reverse=True)

    return images


def delete_image(s3_client, key: str) -> None:
    """Delete an image from the gallery bucket."""
    s3_client.delete_object(Bucket=minio_config.bucket, Key=key)
